### PR TITLE
New swap function

### PR DESCRIPTION
**Pull request by albanD**
This PR is proposing a new approach to solve the nn/optim only linked by python object identity problem.
The idea is to have a function that can swap the content of two Tensors t1 and t2 while preserving all the old references.
This would allow us to swap the `model.weight` with a new Tensor (can be any subclass of Tensor and any TensorImpl (xla, sparse, nested tensorimpl would work)). The use within nn will be done in a follow up.

This is done by swapping the whole content of the PyObject and then putting back the fields associated with external references (refcount, gc tracking and weakrefs).
Note that we have to properly handle all the cases where there is memory used before the public pointer PyObject* and where the PyObject is bigger due to dict/weakref being inlined (older CPython version) or due to slots.

The main limitation of this approach is that the number of slots need to match for the objects being swapped and thus limit usage of slots in subclasses.

Draft right now to see what @colesbury thinks about doing this?

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

**Comment by albanD** [comment]
Failures are interesting. All test pass for local debug builds on 3.8 and 3.11 but release builds seems to fail. I guess object layout is not the same...

File: torch/csrc/utils/python_compat.h [hunk 1]
```diff
@@ -32,6 +32,27 @@ PyCode_GetNFreevars(PyCodeObject* code) {
...
+  uintptr_t _gc_next;
+
+  // Pointer to previous object in the list.
+  // Lowest two bits are used for flags documented later.
+  uintptr_t _gc_prev;
+} PyGC_Head;
+
+/* Get an object's GC head */
+static inline PyGC_Head* _Py_AS_GC(PyObject* op) {
+  char* gc = ((char*)op) - sizeof(PyGC_Head);
```
Comment (malfet) [inline torch/csrc/utils/python_compat.h:51]: Consider using `offsetof` rather than `sizeof` in case compiler decides to do  some padding

**Comment by albanD** [comment]
Feedback from Sam:
- This is going to heavily break in 3.13. And it will continue breaking going forward. Are we ok with the maintenance burden being introduced by this?
- Instead of doing a "copy everything and put back what we didn't want to move" approach, can we do the opposite "only copy what you care about": `__dict__`, the type (not clear what is the best way to do that), slots (forbidden or copy them?), the c++ Tensor object. This is much less likely to keep breaking with new versions and will allow us to re-use builtin type swapping checks which will be more reliable than ours.

I think the answer to 1 is yes if it is the only way for us to allow updating an nn.Module with subclass well.
I will take a stab at doing 2 and see how it compares.

**Comment by albanD** [comment]
@colesbury @ezyang this is ready for final review!

**Comment by ezyang** [comment]
going off to recharge, I'll let @colesbury grab it

File: torch/csrc/dynamo/cpython_defs.h [hunk 2]
```diff
@@ -1,6 +1,8 @@
 #pragma once
 
+#define Py_BUILD_CORE
```
Comment (colesbury) [inline torch/csrc/dynamo/cpython_defs.h:3]: Why is this needed?

File: torch/csrc/utils/python_compat.h [hunk 3]
```diff
@@ -32,6 +32,27 @@ PyCode_GetNFreevars(PyCodeObject* code) {
...
+  // 0 means the object is not tracked
+  uintptr_t _gc_next;
+
+  // Pointer to previous object in the list.
+  // Lowest two bits are used for flags documented later.
+  uintptr_t _gc_prev;
+} PyGC_Head;
+
+/* Get an object's GC head */
+static inline PyGC_Head* _Py_AS_GC(PyObject* op) {
```
Comment (colesbury) [inline torch/csrc/utils/python_compat.h:50]: Why is this needed? It doesn't look like it's used in the latest version of the PR?

File: torch/utils/__init__.py [hunk 4]
```diff
@@ -20,3 +21,32 @@ def set_module(obj, mod):
...
+        setattr(t2, name, tmp)
+
+    # Swap the types
+    # Note that this will fail if there are mismatched slots
+    swap_attr("__class__")
+
+    # Swap the dynamic attributes
+    swap_attr("__dict__")
+
+    # Swap the at::Tensorr they point to
```
Comment (colesbury) [inline torch/utils/__init__.py:51]: Tensorrrrr

File: torch/csrc/utils/python_compat.h (hunk 3 above)
Comment (albanD) [reply torch/csrc/utils/python_compat.h:50]: Not needed anymore! Good catch!

File: torch/csrc/dynamo/cpython_defs.h (hunk 2 above)
Comment (albanD) [reply torch/csrc/dynamo/cpython_defs.h:3]: This was conflicting when defining some GC/weakref structs. But these are not needed anymore!

**Review by colesbury — APPROVED** [review]
